### PR TITLE
Standalone wear 3

### PR DESCRIPTION
**Pull request by kskandispersonal**
I created this branch as following:
`kskandis@Pasha MINGW64 /d/Code/xDrip-plus (standalone-wear-2)
$ git checkout -b standalone-wear-3 standalone-wear-2
Switched to a new branch 'standalone-wear-3'

kskandis@Pasha MINGW64 /d/Code/xDrip-plus (standalone-wear-3)
$ git push origin standalone-wear-3
Total 0 (delta 0), reused 0 (delta 0)
To https://github.com/kskandispersonal/xDrip-plus.git
- [new branch]      standalone-wear-3 -> standalone-wear-3

But for some reason it isn't showing the changes you made in upstream, and doesn't reflect what I see when I do compare between local branch and upstream branch (which only shows a few diffs).

### CODE REVIEW

**Comment by jotomo** [comment]
@kskandispersonal from what you wrote you only created a new branch that's the same as the _wear-2_ branch. Did you miss a _git fetch uppstream && git rebase upstream/master_ perhaps?
@jamorham FYI, if you don't already know: kskandis' fork is has the remote name _origin_ and your repo has the remote name _upstream_.

**Comment by kskandispersonal** [comment]
This looks better!  I followed @jotomo's advice:
```
kskandis@Pasha MINGW64 /d/Code/xDrip-plus (standalone-wear-2)
git checkout -b standalone-wear-3 standalone-wear-2
git push origin standalone-wear-3
$ git push origin standalone-wear-3
--> Git still showed Files Changed: 90 in new PR for this branch
$ git fetch upstream
$ git rebase upstream/master
kskandis@Pasha MINGW64 /d/Code/xDrip-plus (standalone-wear-3|REBASE 6/11)
$ git rebase --continue
$ git rebase --continue
--> Resolve conflicts in AS
$ git reset upstream/master
$ git commit -a -m "Merge server code"
$ git push --force origin standalone-wear-3
--> PR for this branch shows 10 files changed
```

**Comment by kskandispersonal** [comment]
@jotomo Were those git commands correct?  It seems like it finally looks correct in the PR.

File: app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/BgSendQueue.java [hunk 1]
```diff
@@ -108,7 +109,8 @@ public static void handleNewBgReading(BgReading bgReading, String operation_type
                     context.sendBroadcast(updateIntent);
                 }
 
-                if (AppWidgetManager.getInstance(context).getAppWidgetIds(new ComponentName(context, xDripWidget.class)).length > 0) {
+                Context appContext = xdrip.getAppContext();//KS
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/BgSendQueue.java]: I wasn't sure about this bit, is it needed? what was the reason for this change?

**Comment by jamorham** [comment]
I think we are nearly there!! I will finish merging this off tomorrow and then will be asking you to test if it still works!

File: app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/BgSendQueue.java (hunk 1 above)
Comment (kskandispersonal) [inline app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/BgSendQueue.java]: I think I needed this after I started calling BgSendQueue from WatchUpdateService.  For some reason, it didn't work as originally coded.  I will re-test this though, as I am now wondering maybe there could have been some other reason AppWidget was failing...

**Comment by jotomo** [comment]
@jamorham Yes, those git commands look good. Seems we're nearly there indeed :-)

File: wear/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -60,4 +60,25 @@
     <string name="persistent_high_for_greater_than">Persistent High for > </string>
     <string name="space_mins">" mins"</string>
 
+    <string name="location_permission_not_approved">&lt;br&gt;&lt;br&gt;&lt;br&gt;WatchFace requires Location permission. Click on this WatchFace or visit Settings &gt; Permissions to approve.</string>
```
Comment (jamorham) [inline wear/src/main/res/values/strings.xml]: This is already merged, your version has two copies of each of the new strings and I expect wont compile

**Comment by jamorham** [comment]
Manually merged! Phew!

@kskandispersonal okay can you now clone a fresh copy of the current master tree and compile that and tell me does it work in reality?

@jotomo thanks for all your assistance, much appreciated

File: wear/src/main/java/com/eveningoutpost/dexdrip/Home.java [hunk 3]
```diff
@@ -25,6 +25,12 @@
     @Override
     public void onCreate() {
         super.onCreate();
+
+        //KS copied from app/Home
+        Home.context = getApplicationContext();
+        set_is_follower();
```
Comment (kskandispersonal) [inline wear/src/main/java/com/eveningoutpost/dexdrip/Home.java]: Yes, that's right, wear/BgReading is used when wear watch is running in standalone-g5 mode.  Yes, I wanted to keep wear/BgReading as similar to app/BgReading as possible so I reproduced calls in wear/Home for follower.  I think it is safe to set follower to false, which is what I was originally doing:)

File: wear/src/main/res/values/strings.xml (hunk 2 above)
Comment (kskandispersonal) [inline wear/src/main/res/values/strings.xml]: Hmm... I must have made a mistake adding this back in during Resolve Conflict.:(

**Comment by kskandispersonal** [comment]
Cloned, and will test today.

**Comment by kskandispersonal** [comment]
So far, the cloned version is running as expected.  I haven't had a chance to take the watch running in standalone mode by itself yet though, maybe tomorrow.  Also, I need to change my sensor as it seems to be dying:(  Changing the sensor will be a good test, too.  But so far, everything as expected.  I've been mainly running the G5 Collector on the watch and BGs are being uploaded to the phone as received from the transmitter.  More testing should be done with the preferences but it appears to be running as expected:)

**Comment by jamorham** [comment]
Okay this is good news so far! When I tried to test this with my LG watch it took ages (25 minutes) before android wear actually downloaded the new version of the microapk to the watch. I assume this is something you would have experience of during the development but what I want to check is that you are certain that the microapk running on the watch is actually the new one from this cloned code?

I also was getting messages about deadlocks pop up on the phone, can you keep an eye out for any of these? As I don't usually use android wear I am not sure if these are unrelated. I have also been hearing that there may be some other unrelated issues to do with recent updates to the google android wear app itself causing issues.

**Comment by kskandispersonal** [comment]
Wow, no!  I have only been testing with the watch connected to the USB, and then I build the wear debug version which gets installed immediately, less than a minute.  I actually have not tried to build an install.  It sounds like 25 minutes is ridiculously long.

Yes, I created a new folder for the cloned version and that is the one I built and could see it being installed on the watch, the debug version.  I have actually debugged any code though. 

Yes, I also get the wear deadlocks popup on the phone.  I think it is related to code in D:\Code\xDrip-plus\app\src\main\java\com\eveningoutpost\dexdrip\wearintegration\SendToDataLayerThread.java but I don't know why it is occurring.  WatchUpdateService uses it to send the data async to the watch.  It doesn't appear to effect the execution on the watch or phone.  So, yes, it is related to wear.

**Comment by jamorham** [comment]
Yes I think my LG might be a bit buggy, it took 3 hours to sync my contacts when I first got it nothing related to xDrip at all.

Okay I will look at the deadlock stuff as that detection is code I wrote. It might be it is processing more data than before and if requests are overlapping it is triggering that. I haven't started looking at it yet.

Jon

**Comment by kskandispersonal** [comment]
Okay.  I am going to change my sensor tonight as my readings are unstable.  It will be better to test with, too.

**Comment by jotomo** [comment]
Syncing should take be under a minute. In the Android Wear app, in the settings for the watch, there's also a _resync apps_ button force a sync.

**Comment by kskandispersonal** [comment]
Oh, I didn't realize there was a resync apps button on the wear!  I will give that a try.

**Comment by kskandispersonal** [comment]
Hmm, I just tried it.  It was just a few seconds.  How can you tell whether it re-synced or not?!
I just looked at Watch Storage in the Android Wear app on the phone.  It shows XDrip Prefs. using 28Mb.  Google Fit is using 20Mb. So I guess XDrip is acceptable though pretty hefty!

**Comment by jamorham** [comment]
I just added a new commit: https://github.com/jamorham/xDrip-plus/commit/0b22d24f59fe005387a4c4240476aea1863ad088 - There was a null pointer issue but more significantly the G5 service would get started if Wear Integration was used even if the user is using a different data source.  I also added some extra debugging for the deadlock issue.

Once we get a bit further with successful testing there will be a job to clean things up a bit and I think only show G5 menu items to people who are using G5.

**Comment by kskandispersonal** [comment]
Oh, my bad, I overlooked that not all wear users would be G5!:(  Very significant, indeed!:(  I retrieved your changes and am using that build now on my watch.

Yes, I agree, there will be a big job in cleaning up the code and yes, G5 menu items for G5 users.  I guess this is just a first step, sort of a proof of concept!

**Comment by jamorham** [comment]
I think I've fixed the deadlock warnings. I believe they were caused by a race condition resetting the `googleapiclient` reference but I've also adjusted the whole multi-threading in `SendToDataLayerThread.java` with a lock. Please also try out this update.

**Comment by kskandispersonal** [comment]
Just built and installed so far so good.  No wear deadlocks:)

**Comment by jamorham** [comment]
I believe I can close this now as the feature has been merged (thank you so much!) and there are other PRs relating to updates to the code etc.

**Pull request closed without merging**